cli/sdl_generate.go

```diff
@@ -0,0 +1,291 @@
+			}
+
+			cols, err := schemaManager.ParseSDL(string(sdlBuf))
+			if err != nil {
+				return errors.Join(ErrParsingSDL, err)
+			}
+
+			collections := make([]client.CollectionVersion, len(cols))
+			for i, c := range cols {
+				collections[i] = c.Definition
+			}
+
+			cache := description.NewCollectionCache()
+			cache.AddAll(collections)
+			ctx := description.WithCollectionCache(ctx, cache)
+
+			_, err = schemaManager.Generator.Generate(ctx, collections)
+			if err != nil {
+				return errors.Join(ErrGeneratingSDL, err)
+			}
+
+			params := gql.Params{Schema: *schemaManager.Schema(), RequestString: introspectionQueryRequest}
+			r := gql.Do(params)
+			if len(r.Errors) != 0 {
+				// for simplicity we're just going to return the
+				// first error, if there are more, they'll be caught on
+				// follow up invocations.
+				return errors.Join(ErrGeneratingSDL, r.Errors[0])
+			}
+
+			respJson, err := json.Marshal(r.Data)
+			if err != nil {
+				return err
+			}
+			respBuf := bytes.NewBuffer(respJson)
+
+			converter := introspection.JsonConverter{}
+			doc, err := converter.GraphQLDocument(respBuf)
+			if err != nil {
+				return err
+			}
+
+			var outWriter io.Writer
+			if outputFile == "-" {
+				outWriter = os.Stdout
+			} else {
+				// check if the file exists, if so check for the overwrite
+				// flag
+				ofinfo, err := os.Stat(outputFile)
+				if err != nil && !errors.Is(err, os.ErrNotExist) {
+					return err
+				}
+				if ofinfo != nil && !yesOverwrite {
+					fmt.Fprintln(os.Stderr, "output file path already exists. If you want to overwrite use -y")
+					os.Exit(1)
+				}
+
+				f, err := os.OpenFile(outputFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+				if err != nil {
+					return err
+				}
+				defer f.Close() //nolint:errcheck
+				outWriter = f
+			}
+
+			err = astprinter.PrintIndent(doc, []byte("    "), outWriter)
+			if err != nil {
+				return errors.Join(ErrWritingOutput, err)
+			}
+
+			return nil
+		},
+	}
+
+	EmbedCLIExample(ctx, cmd, "Generate SDL",
+		`defradb sdl generate foo.graphql`)
+
+	EmbedCLIExample(ctx, cmd, "Generate Multiple SDLs",
+		`defradb sdl generate foo.graphql bar.graphql`)
+
+	EmbedCLIExample(ctx, cmd, "Generate SDL and overwrite output",
+		`defradb sdl generate foo.graphql bar.graphql --output schema.graphql -y`)
+
+	cmd.PersistentFlags().StringVarP(&outputFile, "output", "o", defaultOutputPath,
+		"The output file to write the generated schema. Accepts '-' to write to stdout")
+
+	EmbedCLIExample(ctx, cmd, "Generate SDL with Searchable Encryption type definitions",
+		`defradb sdl generate foo.graphql -s`)
+
+	cmd.PersistentFlags().BoolVarP(&yesOverwrite, "overwrite", "y", false,
+		"Overwrite any existing matching output file paths")
+
+	cmd.PersistentFlags().BoolVarP(&searchableEncryption, "include-searchable-encryption", "s",
+		false, "Include the schema type definitions to support Searchable Encryption")
+
+	return cmd
+}
+
+var introspectionQueryRequest = `
```

Comment:
suggestion: There are about 1000 whitespace characters in this variable that will take space on the binary (about 1kB). Since this probably wont change over time, it might make sense to inline everything to reduce size. Everything counts if we want to minimize binary size :)